src/realm/node_header.hpp

```diff
@@ -229,16 +301,631 @@ class NodeHeader {
+    auto h = header;
+
+    auto encoding = get_encoding(h);
+    REALM_ASSERT_DEBUG(encoding >= Encoding::WTypBits && encoding <= Encoding::Flex);
+    switch (encoding) {
+        case Encoding::WTypBits:
+        case Encoding::WTypIgn:
+        case Encoding::WTypMult: {
+            WidthType wtype = get_wtype_from_header(header);
+            size_t width;
+            size_t size;
+            width = get_width_from_header(header);
+            size = get_size_from_header(header);
+            return calc_byte_size(wtype, size, static_cast<uint_least8_t>(width));
+        }
+        case Encoding::Packed:
+            return NodeHeader::header_size + align_bits_to8(get_num_elements<NodeHeader::Encoding::Packed>(h) *
+                                                            get_element_size<NodeHeader::Encoding::Packed>(h));
+        case Encoding::AofP:
+            return NodeHeader::header_size + align_bits_to8(get_num_elements<NodeHeader::Encoding::AofP>(h) *
+                                                            (get_elementA_size<NodeHeader::Encoding::AofP>(h) +
+                                                             get_elementB_size<NodeHeader::Encoding::AofP>(h)));
+        case Encoding::PofA:
+            return NodeHeader::header_size + align_bits_to8(get_num_elements<NodeHeader::Encoding::PofA>(h) *
+                                                            (get_elementA_size<NodeHeader::Encoding::PofA>(h) +
+                                                             get_elementB_size<NodeHeader::Encoding::PofA>(h)));
+        case Encoding::Flex:
+            return NodeHeader::header_size + align_bits_to8(get_arrayA_num_elements<NodeHeader::Encoding::Flex>(h) *
+                                                                get_elementA_size<NodeHeader::Encoding::Flex>(h) +
+                                                            get_arrayB_num_elements<NodeHeader::Encoding::Flex>(h) *
+                                                                get_elementB_size<NodeHeader::Encoding::Flex>(h));
+        default:
+            REALM_UNREACHABLE();
+    }
+}
+
+
+uint_least8_t inline NodeHeader::get_width_from_header(const char* header) noexcept
+{
+    REALM_ASSERT_DEBUG(!wtype_is_extended(header));
+    typedef unsigned char uchar;
+    const uchar* h = reinterpret_cast<const uchar*>(header);
+    return uint_least8_t((1 << (int(h[4]) & 0x07)) >> 1);
+}
+
+// A little helper:
+size_t inline NodeHeader::get_size_from_header(const char* header) noexcept
+{
+    using Encoding = NodeHeader::Encoding;
+    if (wtype_is_extended(header)) {
+        const auto encoding = get_encoding(header);
+        if (encoding == Encoding::Flex) {
+            return get_arrayB_num_elements<Encoding::Flex>(header);
+        }
+        else if (encoding == Encoding::Packed) {
+            return get_num_elements<Encoding::Packed>(header);
+        }
+        REALM_UNREACHABLE(); // other encodings are not supported.
```

Comment:
This sounds odd. I guess AodP and PofA sizes can be calculated the same way as Packed

Reply:
Aodp  and PoFa will be deleted...